clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/op/artifact/KubernetesCleanupArtifactsOperation.java

```diff
@@ -0,0 +1,116 @@
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang.StringUtils;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Slf4j
+public class KubernetesCleanupArtifactsOperation implements AtomicOperation<Void> {
+  private final KubernetesCleanupArtifactsDescription description;
+  private final KubernetesV2Credentials credentials;
+  private final String accountName;
+  private final ArtifactProvider artifactProvider;
+  private final KubernetesResourcePropertyRegistry registry;
+  private static final String OP_NAME = "CLEANUP_KUBERNETES_ARTIFACTS";
+
+  public KubernetesCleanupArtifactsOperation(KubernetesCleanupArtifactsDescription description, ArtifactProvider artifactProvider, KubernetesResourcePropertyRegistry registry) {
+    this.description = description;
+    this.credentials = (KubernetesV2Credentials) description.getCredentials().getCredentials();
+    this.accountName = description.getCredentials().getName();
+    this.artifactProvider = artifactProvider;
+    this.registry = registry;
+  }
+
+  private static Task getTask() {
+    return TaskRepository.threadLocalTask.get();
+  }
+
+  @Override
+  public Void operate(List priorOutputs) {
+    List<Artifact> artifacts = description.getManifests().stream()
+        .map(this::artifactsToDelete)
+        .flatMap(Collection::stream)
+        .collect(Collectors.toList());
+
+    artifacts.forEach(a -> {
+      String type = a.getType();
+      if (!type.startsWith("kubernetes/")) {
+        log.warn("Non-kubernetes type deletion requested...");
+        return;
+      }
+      String kind = type.substring("kubernetes/".length());
+      KubernetesResourceProperties properties = registry.get(accountName, KubernetesKind.fromString(kind));
+      if (properties == null) {
+        log.warn("No properties for artifact {}, ignoring", a);
+        return;
+      }
+
+      getTask().updateStatus(OP_NAME, "Deleting artifact '" + a + '"');
+      KubernetesHandler handler = properties.getHandler();
+      String name = a.getName();
+      if (StringUtils.isNotEmpty(a.getVersion())) {
+        name = String.join("-", name, a.getVersion());
+      }
+      // todo add to outputs
+      handler.delete(credentials, a.getLocation(), name, null, new V1DeleteOptions());
+    });
+
+    return null;
+  }
+
+  private List<Artifact> artifactsToDelete(KubernetesManifest manifest) {
+    Artifact artifact = KubernetesManifestAnnotater.getArtifact(manifest);
+    List<Artifact> artifacts = artifactProvider.getArtifacts(artifact.getType(), artifact.getName(), artifact.getLocation())
+        .stream()
+        .filter(a -> a.getMetadata() != null && accountName.equals(a.getMetadata().get("account")))
+        .collect(Collectors.toList());
+
+    KubernetesManifestStrategy strategy = KubernetesManifestAnnotater.getStrategy(manifest);
+    if (strategy.getMaxVersionHistory() == null) {
```

Comment:
Minor nit: It doesn't look like we need to execute lines 98 - 102 if strategy.getMaxVersionHistory() is null.  This code could go at the start of the function to avoid unnecessary work.

Reply:
Good catch